File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        m,n = len(grid), len(grid[0])
        directions = [(1,0),(0,1),(-1,0),(0,-1)]
        visited = set()
        ans = 0

        def out_of_index(i,j):
            return i>=m or j>=n or i<0 or j<0 

        def dfs(i,j,visited):
            visited.add((i,j))

            for x,y in directions:
                new_x, new_y = i+x, j+y
                if out_of_index(new_x,new_y) or grid[new_x][new_y] == "0" or (new_x,new_y) in visited:
                    continue
                
                dfs(new_x,new_y,visited)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1" and (i,j) not in visited:
                    dfs(i,j,visited)
                    ans+=1
        print(len(visited))
        return ans
        

File: 200-number-of-islands/number-of-islands.java
import java.util.*;

class Solution {
    static int M;
    static int N;

    public int numIslands(char[][] grid) {
        System.out.println(Arrays.deepToString(grid));
        int answer = 0;
        M = grid.length;
        N = grid[0].length;

        boolean[][] visited = new boolean[M][N];

        for(int i=0;i<M;i++){
            for(int j=0;j<N;j++){
                if(!visited[i][j] && grid[i][j] == '1'){
                    answer += 1;
                    bfs(i,j,visited,grid);
                }
            }
        }

        return answer;
    }

    private boolean outOfIndex(int i, int j){
        return i<0 || j<0 || i>=M || j>= N;
    }

    private void bfs(int i, int j, boolean[][] visited, char[][] grid){
        if(outOfIndex(i,j) || visited[i][j] || grid[i][j] == '0'){
            return;
        }

        visited[i][j] = true;

        bfs(i+1,j,visited,grid);
        bfs(i-1,j,visited,grid);
        bfs(i,j+1,visited,grid);
        bfs(i,j-1,visited,grid);
    }
}